### PR TITLE
Feature/8025 dcc validation ux/8246 validation result screens

### PR DESCRIPTION
**Pull request by nickguendling**
## Description
Shows the business rule validation results

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8025

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Extensions/SAP_Internal_Dgc_ValueSet+DisplayText.swift [hunk 1]
```diff
@@ -0,0 +1,17 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Extensions/SAP_Internal_Dgc_ValueSet+DisplayText.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Extensions/SAP_Internal_Dgc_ValueSets+ValueSetType.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Extensions/SAP_Internal_Dgc_ValueSets+ValueSetType.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewHeadlineWithImageCell.swift [hunk 3]
```diff
@@ -0,0 +1,71 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewHeadlineWithImageCell.swift:1]:
```suggestion
//
```

**Comment by Ein-Tim** [comment]
@nickguendling 

Hi Nick, is there any way for the community to test the new screens? 
On my Emulator I'm stuck because it seems impossible to add vaccination certificates. 😕

Thanks for all your work. Have a nice day!

File: src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewHeadlineWithImageCell.swift [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	@available(*, unavailable)
+	required init?(coder: NSCoder) {
+		fatalError("init(coder:) has not been implemented")
+	}
+
+	// MARK: - Overrides
+
+	// MARK: - Protocol <#Name#>
+
+	// MARK: - Public
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewHeadlineWithImageCell.swift:25]: Not needed
```suggestion
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/HealthCertificateValidationResultViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,30 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/HealthCertificateValidationResultViewModel.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/DynamicCell+ValidationResult.swift [hunk 6]
```diff
@@ -0,0 +1,37 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/DynamicCell+ValidationResult.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/TechnicalValidationFailedRulesTableViewCell.swift [hunk 7]
```diff
@@ -0,0 +1,90 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/TechnicalValidationFailedRulesTableViewCell.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/ValidationResultCellModel.swift [hunk 8]
```diff
@@ -0,0 +1,223 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/ValidationResultCellModel.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/ValidationResultTableViewCell.swift [hunk 9]
```diff
@@ -0,0 +1,127 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/ValidationResultTableViewCell.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationFailedViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,118 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationFailedViewModel.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateTechnicalValidationFailedViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,65 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateTechnicalValidationFailedViewModel.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationOpenViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,91 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationOpenViewModel.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationPassedViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,72 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationPassedViewModel.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationResultViewController.swift [hunk 14]
```diff
@@ -0,0 +1,96 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationResultViewController.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationResultViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,12 @@
+////
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationResultViewModel.swift:1]:
```suggestion
//
```

File: src/xcode/ENA/ENA/Source/Services/HealthCertificateValidation/MockHealthCertificateValidationService.swift [hunk 16]
```diff
@@ -16,12 +20,12 @@ struct MockHealthCertificateValidationService: HealthCertificateValidationProvid
...
 	)
 
-	var validationResult: Result<HealthCertificateValidationReport, HealthCertificateValidationError> = .success(.validationPassed([]))
+	var validationResult: Result<HealthCertificateValidationReport, HealthCertificateValidationError> = .success(.validationFailed([validationResult1, validationResult2]))
 
 	func onboardedCountries(
 		completion: @escaping (Result<[Country], HealthCertificateValidationOnboardedCountriesError>) -> Void
 	) {
-		DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
+		DispatchQueue.main.asyncAfter(deadline: .now() + 0) {
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificateValidation/MockHealthCertificateValidationService.swift:28]:
```suggestion
		DispatchQueue.main.asyncAfter(deadline: .now()) {
```

File: src/xcode/ENA/ENA/Source/Services/HealthCertificateValidation/MockHealthCertificateValidationService.swift [hunk 17]
```diff
@@ -32,9 +36,47 @@ struct MockHealthCertificateValidationService: HealthCertificateValidationProvid
 		validationClock: Date,
 		completion: @escaping (Result<HealthCertificateValidationReport, HealthCertificateValidationError>) -> Void
 	) {
-		DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
+		DispatchQueue.main.asyncAfter(deadline: .now() + 0) {
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificateValidation/MockHealthCertificateValidationService.swift:39]:
```suggestion
		DispatchQueue.main.asyncAfter(deadline: .now()) {
```

File: src/xcode/ENA/ENA/Source/Views/GradientView.swift [hunk 18]
```diff
@@ -154,6 +162,33 @@ class GradientView: UIView {
...
+			let lightColors = [
+				UIColor(red: 255 / 255, green: 255 / 255, blue: 255 / 255, alpha: 1).cgColor,
+				UIColor(red: 255 / 255, green: 255 / 255, blue: 255 / 255, alpha: 1).cgColor,
+				UIColor(red: 235 / 255, green: 244 / 255, blue: 255 / 255, alpha: 1).cgColor
+			]
+
+			let darkColors = [
+				UIColor(red: 25 / 255, green: 25 / 255, blue: 27 / 255, alpha: 1).cgColor,
+				UIColor(red: 47 / 255, green: 65 / 255, blue: 77 / 255, alpha: 1).cgColor
+			]
```
Comment (whiskey) [inline src/xcode/ENA/ENA/Source/Views/GradientView.swift:176]: Should we define them in the ena colors?

**Review by whiskey — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 19]
```diff
@@ -2863,3 +2863,71 @@ Die Angaben werden statistisch ausgewertet, sie werden nicht zu einem Profil ges
...
+
+"HealthCertificate_Validation_Open_openSectionTitle" = "Hinweis";
+
+"HealthCertificate_Validation_Open_openSectionDescription" = "Nachfolgende Regeln konnten nicht geprüft werden. Vergewissern Sie sich, dass das Zertifikat den Regeln Ihres Reiselandes entspricht.";
+
+/* - Failed */
+
+"HealthCertificate_Validation_Failed_title" = "Zertifikat nicht gültig";
+
+"HealthCertificate_Validation_Failed_subtitle" = "Ihr Zertifikat ist im gewählten Land nicht gültig.";
```
Comment (NataliaLemmerth) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:2913]:
```suggestion
"HealthCertificate_Validation_Failed_subtitle" = "Ihr Zertifikat ist im gewählten Land nicht gültig";
```

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 20]
```diff
@@ -2863,3 +2863,71 @@ Die Angaben werden statistisch ausgewertet, sie werden nicht zu einem Profil ges
...
+
+"HealthCertificate_Validation_Passed_hintsTitle" = "Hinweis";
+
+"HealthCertificate_Validation_Passed_hint1" = "Beachten Sie, dass sich die Einreiseregeln ändern können. Prüfen Sie daher das Zertifikat kurz vor der Einreise (max. 48 Stunden).";
+
+"HealthCertificate_Validation_Passed_hint2" = "Es können in einzelnen Regionen weitere Regeln oder Einschränkungen gelten.";
+
+"HealthCertificate_Validation_Passed_hint3" = "Um die Echtheit eines Zertifikats sicherzustellen, wird jedes Zertifikat mit einer digitalen Signatur ausgestellt. Diese Signatur wird in einer Prüf-Anwendung validiert.";
+
+"HealthCertificate_Validation_Passed_hint4" = "Ob die im Zertifikat eingetragenen Daten richtig sind, wird nicht geprüft.";
```
Comment (NataliaLemmerth) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:2895]:
```suggestion
"HealthCertificate_Validation_Passed_hint4" = "Ob die im Zertifikat eingetragenen Daten richtig sind, wird in der Corona-Warn-App nicht geprüft.";
```

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 21]
```diff
@@ -2863,3 +2863,71 @@ Die Angaben werden statistisch ausgewertet, sie werden nicht zu einem Profil ges
...
+
+"HealthCertificate_Validation_Result_moreInformation02" = "FAQ";
+
+"HealthCertificate_Validation_Result_moreInformation03" = "und unter";
+
+/* - Passed */
+
+"HealthCertificate_Validation_Passed_title" = "Gültiges Zertifikat";
+
+"HealthCertificate_Validation_Passed_subtitle" = "Ihr Zertifikat ist im gewählten Land gültig.";
```
Comment (NataliaLemmerth) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:2885]:
```suggestion
"HealthCertificate_Validation_Passed_subtitle" = "Ihr Zertifikat ist im gewählten Land gültig";
```

File: src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings [hunk 22]
```diff
@@ -2863,3 +2863,71 @@ Die Angaben werden statistisch ausgewertet, sie werden nicht zu einem Profil ges
...
+
+"HealthCertificate_Validation_Failed_openSectionTitle" = "Prüfung nicht möglich";
+
+"HealthCertificate_Validation_Failed_openSectionDescription" = "Nachfolgende Regeln konnten nicht geprüft werden. Vergewissern Sie sich, dass das Zertifikat den Regeln Ihres Reiselandes entspricht.";
+
+/* - Technical Validation Failed */
+
+"HealthCertificate_Validation_TechnicalFailed_title" = "Zertifikat nicht gültig";
+
+"HealthCertificate_Validation_TechnicalFailed_subtitle" = "Ihr Zertifikat ist im gewählten Land nicht gültig.";
```
Comment (NataliaLemmerth) [inline src/xcode/ENA/ENA/Resources/Localization/de.lproj/Localizable.strings:2927]:
```suggestion
"HealthCertificate_Validation_TechnicalFailed_subtitle" = "Ihr Zertifikat ist im gewählten Land nicht gültig";
```

**Review by NataliaLemmerth — CHANGES_REQUESTED** [review]
Some minor text changes according to the actual state of Figma.

File: src/xcode/ENA/ENA/Source/Views/GradientView.swift (hunk 18 above)
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Views/GradientView.swift:176]: Probably not worth it as these are one-off colors just for the gradients

File: src/xcode/ENA/ENA/Source/Services/HealthCertificateValidation/MockHealthCertificateValidationService.swift (hunk 16 above)
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificateValidation/MockHealthCertificateValidationService.swift]: This reminded me to clean up the mocks. I added this initially to check the loading state in the UI. The dispatch to the main queue should not be needed anymore, so I removed it and did some other cleanup.

File: src/xcode/ENA/ENA/Source/Scenes/Checkin/ConfigureableCell.swift [hunk 23]
```diff
@@ -22,4 +22,8 @@ extension ReuseIdentifierProviding {
 		String(describing: self)
 	}
 
+	static var dynamicTableViewCellReuseIdentifier: DynamicTableViewCellReuseIdentifier {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/Checkin/ConfigureableCell.swift:25]: 🙌

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/ValidationResultTableViewCell.swift [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+
+				keyValueAttributedLabel.leadingAnchor.constraint(equalTo: ruleDescriptionLabel.leadingAnchor),
+				keyValueAttributedLabel.topAnchor.constraint(equalTo: ruleTypeDescriptionLabel.bottomAnchor, constant: 16.0),
+				keyValueAttributedLabel.trailingAnchor.constraint(equalTo: ruleDescriptionLabel.trailingAnchor),
+				keyValueAttributedLabel.bottomAnchor.constraint(equalTo: backgroundContainerView.bottomAnchor, constant: -16.0)
+			]
+		)
+	}
+
+	private func updateBorderColor() {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/ValidationResultTableViewCell.swift:122]: Why update if there is no change based on some dynamic state?

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationFailedViewModel.swift [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+			.sorted { $0.rule?.identifier ?? "" < $1.rule?.identifier ?? "" }
+	}
+
+	private var openValidationResults: [ValidationResult] {
+		openAcceptanceRuleValidationResults + openInvalidationRuleValidationResults
+	}
+
+	private var openAcceptanceRuleValidationResults: [ValidationResult] {
+		validationResults
+			.filter { $0.rule?.ruleType == .acceptence && $0.result == .open }
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/HealthCertificateValidationFailedViewModel.swift:108]:
```suggestion
			.filter { ($0.rule?.ruleType == .acceptence || $0.rule?.ruleType == .invalidation) && $0.result == .open }
```

And remove openInvalidationRuleValidationResults

**Review by KaiTeuber — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Views/GradientView.swift [hunk 26]
```diff
@@ -46,6 +51,7 @@ class GradientView: UIView {
 		case lightBlue(withStars: Bool)
 		case mediumBlue(withStars: Bool)
 		case darkBlue(withStars: Bool)
+		case certificateCheck
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Views/GradientView.swift:54]: Naming does not fit to the other types.

**Review by ArturFriesen — CHANGES_REQUESTED** [review]

**Review by whiskey — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewHeadlineWithImageCell.swift [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+	override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+		super.init(style: style, reuseIdentifier: reuseIdentifier)
+		setupView()
+	}
+
+	@available(*, unavailable)
+	required init?(coder: NSCoder) {
+		fatalError("init(coder:) has not been implemented")
+	}
+
```
[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/DynamicTableViewController/Views/DynamicTableViewHeadlineWithImageCell.swift:20]: Removed

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/HealthCertificateValidationResultViewModel.swift [hunk 28]
```diff
@@ -0,0 +1,30 @@
+//
+// 🦠 Corona-Warn-App
+//
+
+import Foundation
+import UIKit
+
+struct HealthCertificateValidationResultViewModel {
```
[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/HealthCertificateValidationResultViewModel.swift]: Ah, that was just an old file lying around. Good find! Removed it :)

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/ValidationResultTableViewCell.swift [hunk 29]
```diff
@@ -0,0 +1,126 @@
...
+		// only needed if keyValuePairs get updated while the cell is shown.
+		// this is why we drop the first callback
+		cellModel.$keyValuePairs
+			.dropFirst()
+			.receive(on: DispatchQueue.main.ocombine)
+			.sink { [weak self] _ in
+				self?.keyValueAttributedLabel.attributedText = cellModel.keyValueAttributedString
+				onUpdate()
+			}
+			.store(in: &subscriptions)
```
[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/ValidationResultTableViewCell.swift:55]: Oh yes, we need to retain the cellModel here and throw it and all the subscriptions away on reuse. Added that now

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/ValidationResultTableViewCell.swift [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+	// MARK: - Private
+
+	private let backgroundContainerView = UIView()
+	private let iconImageView = UIImageView()
+	private let ruleDescriptionLabel = ENALabel(style: .body)
+	private let ruleTypeDescriptionLabel = ENALabel(style: .footnote)
+	private let keyValueAttributedLabel = UILabel()
+	private var subscriptions = Set<AnyCancellable>()
+
+	private func setupView() {
```
[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/Validation/Result/Cells/ValidationResultTableViewCell.swift]: Done 👍

File: src/xcode/ENA/ENA/Source/Views/GradientView.swift (hunk 26 above)
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Views/GradientView.swift]: Renamed it to take the context out and just reflect the colors as we did for the others

**Review by ArturFriesen — APPROVED** [review]

**Review by NataliaLemmerth — APPROVED** [review]
UA approved

**Pull request merged**